arbor/network_impl.cpp

```diff
@@ -0,0 +1,388 @@
+    std::vector<char> labels;
+    std::unordered_map<std::string, cell_gid_type> label_map;
+
+    site_mapping() = default;
+
+    inline std::size_t size() const { return sites.size(); }
+
+    void insert(const network_full_site_info& s) {
+        const auto insert_pair = label_map.insert({std::string(s.label), labels.size()});
+        // append label if not contained in labels
+        if (insert_pair.second) {
+            labels.insert(labels.end(), s.label.begin(), s.label.end());
+            labels.push_back('\0');
+        }
+        sites.emplace_back(distributed_site_info{s.gid,
+            s.lid,
+            s.kind,
+            insert_pair.first->second,
+            s.location,
+            s.global_location,
+            s.hash});
+    }
+
+    network_full_site_info get_site(std::size_t idx) const {
+        const auto& s = this->sites.at(idx);
+
+        network_full_site_info info;
+        info.gid = s.gid;
+        info.lid = s.lid;
+        info.kind = s.kind;
+        info.label = labels.data() + s.label_start_idx;
+        info.location = s.location;
+        info.global_location = s.global_location;
+        info.hash = s.hash;
+
+        return info;
+    }
+};
+
+void push_back(std::vector<connection>& vec,
+    const network_full_site_info& src,
+    const network_full_site_info& dest,
+    double weight,
+    double delay) {
+    vec.emplace_back(connection{{src.gid, src.lid}, {dest.gid, dest.lid}, weight, delay});
+}
+
+void push_back(std::vector<network_connection_info>& vec,
+    const network_full_site_info& src,
+    const network_full_site_info& dest,
+    double weight,
+    double delay) {
+    vec.emplace_back(network_connection_info{
+        network_site_info{
+            src.gid, src.kind, std::string(src.label), src.location, src.global_location},
+        network_site_info{
+            dest.gid, dest.kind, std::string(dest.label), dest.location, dest.global_location},
+        weight,
+        delay});
+}
+
+template <typename ConnectionType>
+std::vector<ConnectionType> generate_network_connections(const recipe& rec,
+    const context& ctx,
+    const domain_decomposition& dom_dec) {
+    const auto description_opt = rec.network_description();
+    if (!description_opt.has_value()) return {};
+
+    const distributed_context& distributed = *(ctx->distributed);
+
+    const auto& description = description_opt.value();
+
+    site_mapping src_sites, dest_sites;
+    std::mutex src_sites_mutex, dest_sites_mutex;
+
+    const auto selection_ptr = thingify(description.selection, description.dict);
+    const auto weight_ptr = thingify(description.weight, description.dict);
+    const auto delay_ptr = thingify(description.delay, description.dict);
+
+    const auto& selection = *selection_ptr;
+    const auto& weight = *weight_ptr;
+    const auto& delay = *delay_ptr;
+
+    std::unordered_map<cell_kind, std::vector<cell_gid_type>> gids_by_kind;
+
+    for (const auto& group: dom_dec.groups()) {
```

Comment:
Technically, `partition_load_balance` already produces gids partitioned
by group and kind. Maybe this could be formalised and encode by an index in the `domain_decomposition` struct?